modeshape-jcr/src/main/java/org/modeshape/jcr/spi/index/provider/IndexChangeAdapters.java

```diff
@@ -334,26 +439,35 @@ protected void removeNode( String workspaceName,
                                    Path path,
                                    Name primaryType,
                                    Set<Name> mixinTypes ) {
-            if (includeRoot || path.isRoot()) {
-                index.remove(nodeKey(key));
-            }
+            index().remove(nodeKey(key));
         }
 
         @Override
-        protected void completeChanges() {
-            index.commit();
-            super.completeChanges();
-        }
-
-        @Override
-        protected void completeWorkspaceChanges() {
-            index.commit();
-            super.completeWorkspaceChanges();
+        protected void renameNode( String workspaceName, 
+                                   NodeKey key, 
+                                   Path newPath, 
+                                   Path.Segment oldSegment, 
+                                   Name primaryType,
+                                   Set<Name> mixinTypes ) {
+            PathFactory pathFactory = context.getValueFactories().getPathFactory();
+            Path oldPath = pathFactory.create(newPath.subpath(0, newPath.size()), oldSegment);
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
         }
 
         @Override
-        public String toString() {
-            return getClass().getSimpleName() + "(\"" + index.getName() + "\")";
+        protected void reorderNode( String workspaceName, 
+                                    NodeKey key, 
+                                    Name primaryType, 
+                                    Set<Name> mixinTypes, 
+                                    NodeKey parent,
+                                    Path newPath, 
+                                    Path oldPath, 
+                                    Path reorderedBeforePath ) {
+            String nodeKey = nodeKey(key);
+            index().remove(nodeKey, valueOf(oldPath));
+            index().add(nodeKey, valueOf(newPath));
```

Comment:
For nodes that do not use SNS, then you are correct in that nothing about child nodes after the `reorderedBeforePath` need be changed.

However, if SNS nodes are being reordered, then reordering changes the SNS index of the node being reordered *and* those nodes after `reorderedBeforePath`. I can give an example if that'd help.